pkg/util/utils.go

```diff
@@ -354,3 +360,185 @@ func OpenExclusiveFile(path string) (*os.File, error) {
+		}
+
+	}
+	return u, err
+}
+
+// StreamFileToStdout streams a tar archive from srcPath to STDOUT
+func StreamFileToStdout(srcPath string, srcfi os.FileInfo) error {
+	if srcfi.IsDir() {
+		tw := tar.NewWriter(os.Stdout)
+		defer tw.Close()
+		err := filepath.Walk(srcPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil || !info.Mode().IsRegular() || path == srcPath {
+				return err
+			}
+			hdr, err := tar.FileInfoHeader(info, "")
+			if err != nil {
+				return err
+			}
+
+			if err = tw.WriteHeader(hdr); err != nil {
+				return err
+			}
+			fh, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			defer fh.Close()
+
+			_, err = io.Copy(tw, fh)
+			return err
+		})
+		if err != nil {
+			return errors.Wrapf(err, "error streaming directory %s to Stdout", srcPath)
+		}
+		return nil
+	}
+
+	file, err := os.Open(srcPath)
+	if err != nil {
+		return errors.Wrapf(err, "error opening file %s", srcPath)
+	}
+	defer file.Close()
+	if !archive.IsArchivePath(srcPath) {
+		tw := tar.NewWriter(os.Stdout)
+		defer tw.Close()
+		hdr, err := tar.FileInfoHeader(srcfi, "")
+		if err != nil {
+			return err
+		}
+		err = tw.WriteHeader(hdr)
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(tw, file)
+		if err != nil {
+			return errors.Wrapf(err, "error streaming archive %s to Stdout", srcPath)
+		}
+		return nil
+	}
+
+	_, err = io.Copy(os.Stdout, file)
+	if err != nil {
+		return errors.Wrapf(err, "error streaming file to Stdout")
+	}
+	return nil
+}
+
+// CopyRemote copies files(transferred by varlink) from srcFilePath to cleanDestPath.
+func CopyRemote(srcFilename string, srcFromInput string, srcFilePath string, destFromInput string, cleanDestPath string, srcIsDir bool, chownOpts *idtools.IDPair, untarPath, copyFileWithTar func(src, dest string) error, extract bool) (string, error) {
+	destdir := cleanDestPath
+	if !srcIsDir && !strings.HasSuffix(destFromInput, string(os.PathSeparator)) {
+		destdir = filepath.Dir(cleanDestPath)
+	}
+	_, err := os.Stat(destdir)
+	if err != nil && !os.IsNotExist(err) {
+		return "", errors.Wrapf(err, "error checking directory %q", destdir)
+	}
+	// needs destDirIsExist to generate destination path
+	destDirIsExist := (err == nil)
+
+	if err = os.MkdirAll(destdir, 0755); err != nil {
+		return "", errors.Wrapf(err, "error creating directory %q", destdir)
+	}
+
+	if srcIsDir {
+		logrus.Debugf("srcFilePath: %s", srcFilePath)
+		destPath := cleanDestPath
+		if destDirIsExist && !strings.HasSuffix(srcFromInput, fmt.Sprintf("%s.", string(os.PathSeparator))) {
+			destPath = filepath.Join(destPath, srcFilename)
+		}
+		logrus.Debugf("copy directory %s to %s", srcFilePath, destPath)
+		if err := untarPath(srcFilePath, destPath); err != nil {
+			return "", errors.Wrapf(err, "error untarring %q to %q", srcFilePath, destPath)
+		}
+		return destPath, nil
+	}
+
+	// untar tempFile under os.TempDir() if the source is a file and not archive
+	tempTarball, err := os.Open(srcFilePath)
```

Comment:
Does the tarball carry the modes for the source files and directory path?  If so that could clean up the API a bit.

Reply:
I'm not sure. I'll check. But the mode argument is only useful when using CopyTo() to copy to remote. It is not useful in the CopyFrom or the varlink call